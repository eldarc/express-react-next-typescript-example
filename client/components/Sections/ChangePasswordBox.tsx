import * as React from 'react'
import { observer, inject } from 'mobx-react'
import Router from 'next/router'

// ~~~ Layout ~~~
import { Button, Form, FormGroup, Input, Label, Alert, Spinner } from 'reactstrap'

// ~~~ Interfaces ~~~
interface IProps {
  [x: string]: any
}

interface IState {
  processing: boolean
  error?: string
}

// ~~~ Component ~~~
@inject('store')
@observer
class SigninBox extends React.Component<IProps, IState> {
  private readonly currentPassword: React.RefObject<HTMLInputElement>
  private readonly newPassword: React.RefObject<HTMLInputElement>

  constructor(props: IProps) {
    super(props)

    this.currentPassword = React.createRef()
    this.newPassword = React.createRef()

    this.state = {
      processing: false,
      error: undefined
    }
  }

  public changePassword = async (event: React.FormEvent) => {
    event.preventDefault()

    try {
      await this.setState({
        processing: true
      })

      // @ts-ignore
      await this.props.store.auth.doChangePassword(this.currentPassword.current.value, this.newPassword.current.value)
      await Router.push('/overview')
    } catch (e) {
      this.setState({
        error: JSON.stringify(e.response.data)
      })
    } finally {
      this.setState({
        processing: false
      })
    }
  }

  public render() {
    return (
        <div className="change-password-box w-100">
          <div className="row">
            <div className="col-12 text-center mb-3">
              <h1>Change your password</h1>
            </div>
            <div className="col-12">
              { this.state.error ? (
                  <Alert color="danger">
                    There was an error while signing in: {this.state.error}
                  </Alert>
              ) : null}
              <Form onSubmit={this.changePassword}>
                <FormGroup>
                  <Label for="currentPassword">Current password</Label>
                  <Input innerRef={this.currentPassword} type="password" name="currentPassword" id="currentPassword" placeholder="Please enter your current password" />
                </FormGroup>
                <FormGroup>
                  <Label for="newPassword">New password</Label>
                  <Input innerRef={this.newPassword} type="password" name="newPassword" id="newPassword" placeholder="Please enter the new password" />
                </FormGroup>

                <div className="d-flex w-100 justify-content-center">
                  <Button type="submit" color="primary">
                    {
                      this.state.processing ? <Spinner size="sm" color="light" />
                      : 'Change password'
                    }
                  </Button>
                </div>
              </Form>
            </div>
          </div>

          {/*language=SCSS*/}
          <style jsx>{`
              @import "variables";

              .change-password-box {
                max-width: 450px;
                background-color: rgba($white, 0.4);
                border-radius: $border-radius-lg;
                padding: $grid-gutter-width / 2;
                box-shadow: 0 1px 2px 0 rgba(60,64,67,0.302),0 1px 3px 1px rgba(60,64,67,0.149);
              }
            `}</style>
        </div>
    )
  }
}

export default SigninBox
