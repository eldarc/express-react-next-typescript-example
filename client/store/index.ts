import { useStaticRendering } from 'mobx-react'
import AuthStore from './modules/AuthStore'
import UserStore from './modules/UserStore'
import OverviewStore from './modules/OverviewStore'
import Cookies from 'js-cookie'

// For SSR static rendering is required.
const isServer = (typeof window === 'undefined')
useStaticRendering(isServer)

// Main root store.
class Store {
  // @ts-ignore (AuthStore is definitely set in the `else` branch in the constructor)
  public auth: AuthStore
  public user: UserStore
  public overview: OverviewStore

  constructor(checkAuth: boolean) {
    this.overview = new OverviewStore()
    this.user = new UserStore()

    // Check if the user is already logged in, and create the store according to that status.
    const token = Cookies.get('token')
    if (checkAuth && token) {
      this.auth = new AuthStore(token)

      // Fetch user data.
      this.user.fetchUserData().then(() => {
        this.overview.ignoreUser(this.user.userData._id)
      })
    } else {
      this.auth = new AuthStore()
    }
  }
}

// Export Store and store initialization function.
export { Store }
let store: Store
export function initializeStore() {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    // return new Store(initialData)
    return new Store(false)
  }

  if (!store) {
    store = new Store(true)
  }

  return store
}
