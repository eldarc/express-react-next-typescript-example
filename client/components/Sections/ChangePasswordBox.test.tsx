/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { Store } from '@/store'

import ChangePasswordBox from './ChangePasswordBox'

describe('ChangePasswordBox', () => {
  it('renders correctly', () => {
    const tree = renderer
        .create(<ChangePasswordBox store={new Store(false)}/>)
        .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
