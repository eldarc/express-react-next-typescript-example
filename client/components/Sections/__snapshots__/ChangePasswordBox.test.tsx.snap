// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordBox renders correctly 1`] = `
<div
  className="jsx-3727646316 change-password-box w-100"
>
  <div
    className="jsx-3727646316 row"
  >
    <div
      className="jsx-3727646316 col-12 text-center mb-3"
    >
      <h1
        className="jsx-3727646316"
      >
        Change your password
      </h1>
    </div>
    <div
      className="jsx-3727646316 col-12"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className=""
            htmlFor="currentPassword"
          >
            Current password
          </label>
          <input
            className="form-control"
            id="currentPassword"
            name="currentPassword"
            placeholder="Please enter your current password"
            type="password"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
            htmlFor="newPassword"
          >
            New password
          </label>
          <input
            className="form-control"
            id="newPassword"
            name="newPassword"
            placeholder="Please enter the new password"
            type="password"
          />
        </div>
        <div
          className="jsx-3727646316 d-flex w-100 justify-content-center"
        >
          <button
            aria-label={null}
            className="btn btn-primary"
            onClick={[Function]}
            type="submit"
          >
            Change password
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
